### ISSUE TITLE
Adjust auction timeout for bidders & latency between servers 

### ISSUE THREAD

**Issue by dbemiller** [prebid/prebid-server#592]
# Motivation

Decrease the number of timeouts from all Bidders.

## Problem Description

Clients set a timeout in OpenRTB through `request.tmax`. Today, Prebid Server simply forwards that value on to each Bidder, along with all the other fields in the ORTB request.

There are two problems with this:

1. Prebid Server adds some overhead by fetching Stored Requests, adding User Sync IDs, and other request preprocessing.
2. There's some latency between the servers hosting PBS and the servers hosting each demand partner. This may vary from Bidder to Bidder, region to region, and even one minute to the next, if the Bidder happens to be under high load.

Since we send them the same timeout, a Bidder who reads the request will think they have more time to respond than they _actually_ do. This might cause them to timeout in the auction without even realizing it.

## Proposal

For (1), Prebid Server can simply time itself during request preprocessing and adjust the timeout before it forwards the request to each Bidder.

For (2), I think we should add some basic machine learning. PBS could run a background thread which periodically pings the endpoint for each Bidder to measure the latency, and use that to estimate how much time is lost on the network. Then we could adjust the `request.tmax` on a per-Bidder basis to account for it.

## Priority

In our (Appnexus') metrics, we don't see too many timeouts from bidders.... so this is fairly low priority for us. However, I wanted to open the ticket because it's a theoretical issue, and may vary from Host to Host. If someone was having issues with bidder timeouts and wanted to work on this, PRs would be welcome.

**Comment by ehoch** [comment]
@dbemiller If this helps you guys prioritize, we're running on the AppNexus Server seeing a fairly large % of timeouts (as defined by the client side Prebid.js timeout event). Higher than when we were client side.

We're busy experimenting with timeouts on both the client and server side, but still seeing significantly higher timeouts than we'd expect, especially at lower server side timeouts (<= 500ms), where we'd really like to see the timeout in a S2S environment. We're talking 30%+ timeouts.

**Comment by dbemiller** [comment]
That's interesting... @ehoch just to confirm, are you using `/auction` or `/openrtb2/auction` as your s2s endpoint?

**Comment by mercuryyy** [comment]
Seeing the same issue here
Using /openrtb2/auction
1.15
all

**Comment by ehoch** [comment]
1. openrtb2
2. 1.15
3. Multiple partners including appnexus
4. We leave a minimum of a 600ms buffer between s2s and client timeouts to leave plenty of time.

**Comment by dbemiller** [comment]
Thanks for the quick responses. I'm not sure what the problem is... but I will share our grafana graphs for per-adapter timeouts, as measured by the Server.

<img width="1676" alt="screen shot 2018-06-28 at 10 46 28 am" src="https://user-images.githubusercontent.com/27972385/42042167-8f191760-7ac1-11e8-9c64-3f063db9e405.png">

If anyone wants to help double-check the code for bugs, the `/openrtb2/auction` endpoint logs adapter requests [here](https://github.com/prebid/prebid-server/blob/master/exchange/exchange.go#L159), and chooses how they get into influx [here](https://github.com/prebid/prebid-server/blob/master/pbsmetrics/go_metrics.go#L329). The graph above was generated from `prebidserver.adapter.{bidder}.requests.timeout.meter`.


This is serving hundreds of `/openrtb2/auction` and `openrtb2/amp` requests per second, so... the server itself is measuring extremely low timeout rates.

I don't know exactly how Prebid.js is measuring s2s timeouts, or what else is going on on the page... but that would be my main suspect atm.

**Comment by ehoch** [comment]
@hhhjort We're padding an extra 600ms on top of what we set Prebid Server timeout to in our tests to make sure the timeout isn't happening on the client roundtrip. Unless we set the Prebid Server timeout to something insanely high (like 1000ms), we're seeing 30%+ timeouts. 1000ms on a S2S timeout kind of defeats the purpose of S2S and will increase our latency. Since that's your default, is it possible the data is skewed? Can we take this offline and use our account # to pull reports specific to our account on Prebid Server?
